staging/src/k8s.io/apimachinery/pkg/api/validation/objectmeta.go

```diff
@@ -130,8 +131,47 @@ func ValidateNoNewFinalizers(newFinalizers []string, oldFinalizers []string, fld
 // ValidateImmutableField validates the new value and the old value are deeply equal.
 func ValidateImmutableField(newVal, oldVal interface{}, fldPath *field.Path) field.ErrorList {
 	allErrs := field.ErrorList{}
+	defer func() {
+		if r := recover(); r != nil {
+			// If we ever get here, then either DeepEqual or cmp.Diff failed.
+			// The former is unlikely and indicates a serious problem with the data,
+			// the latter only gets called when we already know that there is a difference.
+			// In both cases we reject the new value.
+			allErrs = append(allErrs, field.Invalid(fldPath, newVal, FieldImmutableErrorMsg))
+		}
+	}()
+	// We don't do this with cmp.Diff because it is only meant for testing.
+	// It might have bugs that make it less reliable than DeepEqual.
+	if !apiequality.Semantic.DeepEqual(oldVal, newVal) {
+		// However, producing a good error message in those unlikely cases
+		// where a problem was detected is fine and useful, with the defer
+		// above guarding against unexpected panics.
+
+		diff := cmp.Diff(oldVal, newVal)
+		allErrs = append(allErrs, field.Invalid(fldPath, newVal, FieldImmutableErrorMsg+"\n"+diff))
+
+	}
+	return allErrs
+}
+
+// ValidateImmutableFieldWithErr validates the new value and the old value are deeply equal.
+// It accepts an err *field.Error parameter as base error. If the result is not equal, ValidateImmutableFieldWithErr will append additional error information
+// in err.detail and return this error.
+func ValidateImmutableFieldWithErr(newVal, oldVal interface{}, err *field.Error) field.ErrorList {
+	if err == nil {
```

Comment:
I wouldn't special case this, people shouldn't call this method with `err == nil` since there is the other method. On the other hand, I have a feeling that we can indeed implement one function using the other. Can't we do:
```golang
func ValidateImmutableField(newVal, oldVal interface{}, fldPath *field.Path) field.ErrorList {
    return ValidateImmutableFieldWithErr(newVal, oldVal, field.Invalid(fldPath, newVal, FieldImmutableErrorMsg))
}
```

And actually, we can see that you have two implementations, basically the same, one that is nicely documented and one that isn't, removing the duplication will be great, thanks!